### PR TITLE
chore: support old tunnel URLs and clients

### LINKED ISSUE #2: Support old tunnel format
The old tunnel format is a 32 character base16 hostname => 16 byte IPv6 address.

Return both hostnames from the API when registering, with the New And Improved shorter format as the "preferred" URL, but still accept connections from the old hostname.

The old hostname will have the unused bytes from our shorter addresses discarded while being processed.

### PR DESCRIPTION
**Pull request by deansheather**
Changes the register endpoint to return an array of URL strings instead of a single string.

## TODO:
- [x] Add a compatibility endpoint at the same path as the old wgtunnel for non-upgraded clients
- [x] Tests for whether the new URL actually works when accessing the tunnel
- [x] Add specific tests for test cases generated from the old source code just to be double sure
- [x] Add a field to the register endpoint request body to denote which version URL you want to be returned as the primary URL

Closes #2

### CODE REVIEW

File: tunneld/options.go [hunk 1]
```diff
@@ -99,59 +100,106 @@ func (options *Options) Validate() error {
...
+	goodFormatURL.Host = strings.ToLower(goodFormat) + "." + goodFormatURL.Host
+
+	// Old format:
+	oldFormatBytes := make([]byte, 16)
+	copy(oldFormatBytes, addrBytes[:])
+	prefixLenBytes := options.WireguardNetworkPrefix.Bits() / 8
+	copy(oldFormatBytes[prefixLenBytes:], keyHash[:16-prefixLenBytes])
+	oldFormat := hex.EncodeToString(oldFormatBytes)
+	oldFormatURL := *options.BaseURL
+	oldFormatURL.Host = strings.ToLower(oldFormat) + "." + oldFormatURL.Host
```
Comment (mafredri) [inline tunneld/options.go:155]: This "old format" seems different than the previous (old old format), but I suppose this is ok?

File: tunneld/api.go [hunk 2]
```diff
@@ -32,6 +36,7 @@ func (api *API) Router() chi.Router {
 		httpmw.RateLimit(10, 10*time.Second),
 	)
 
+	r.Post("/tun", api.postTun)
 	r.Post("/api/v1/clients", api.postClients)
```
Comment (mafredri) [inline tunneld/api.go:40]: Should we consider introducing a new v2 endpoint?

**Review by mafredri — APPROVED** [review]
Given an occular look, these changes look good. (Haven't actually tested the changes.)

I did notice a data race whilst running the tests via `go test ./... -race -timeout=1m -count=10`, however, it is not necessarily a problem with this implementation:

```
==================
WARNING: DATA RACE
Write at 0x00c004d48490 by goroutine 49358:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  golang.zx2c4.com/wireguard/tun/netstack.(*netTun).Close()
      /home/mafredri/.local/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20230223181233-21636207a675/tun/netstack/tun.go:168 +0xde
  golang.zx2c4.com/wireguard/device.(*Device).Close()
      /home/mafredri/.local/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20230223181233-21636207a675/device/device.go:363 +0x181
  github.com/coder/wgtunnel/tunneld.(*API).Close()
      /home/mafredri/src/coder/wgtunnel/tunneld/tunneld.go:78 +0x64
  github.com/coder/wgtunnel/tunneld_test.createTestTunneldNoDefaults.func1()
      /home/mafredri/src/coder/wgtunnel/tunneld/tunneld_test.go:362 +0x30
  testing.(*common).Cleanup.func1()
      /usr/local/go/src/testing/testing.go:1150 +0x193
  testing.(*common).runCleanup()
      /usr/local/go/src/testing/testing.go:1328 +0x1e9
  testing.tRunner.func2()
      /usr/local/go/src/testing/testing.go:1570 +0x52
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c004d48490 by goroutine 50059:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:160 +0x0
  golang.zx2c4.com/wireguard/tun/netstack.(*netTun).WriteNotify()
      /home/mafredri/.local/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20230223181233-21636207a675/tun/netstack/tun.go:151 +0xe6
  gvisor.dev/gvisor/pkg/tcpip/link/channel.(*queue).Write()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/link/channel/channel.go:100 +0x183
  gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).WritePackets()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/link/channel/channel.go:245 +0xb7
  gvisor.dev/gvisor/pkg/tcpip/stack.(*delegatingQueueingDiscipline).WritePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/stack/nic.go:145 +0x97
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).writeRawPacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/stack/nic.go:393 +0x62
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).writePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/stack/nic.go:389 +0x59
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).WritePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/stack/nic.go:350 +0x22b
  gvisor.dev/gvisor/pkg/tcpip/network/ipv6.(*endpoint).writePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/network/ipv6/ipv6.go:863 +0x408
  gvisor.dev/gvisor/pkg/tcpip/network/ipv6.(*endpoint).WritePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/network/ipv6/ipv6.go:814 +0x2d7
  gvisor.dev/gvisor/pkg/tcpip/stack.(*Route).WritePacket()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/stack/route.go:499 +0xf8
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.sendTCP()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/connect.go:912 +0x3fb
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).sendTCP()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/connect.go:810 +0x199
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).sendRaw()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/connect.go:979 +0x53a
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*sender).sendSegmentFromPacketBuffer()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/snd.go:1679 +0x26d
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*sender).sendSegment()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/snd.go:1640 +0x386
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*sender).maybeSendSegment()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/snd.go:870 +0x704
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*sender).sendData()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/snd.go:974 +0x4fa
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).sendData()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/connect.go:1003 +0xc4
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).shutdownLocked()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/endpoint.go:2535 +0x44f
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).closeLocked()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/endpoint.go:1062 +0xa9
  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).Close()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/transport/tcp/endpoint.go:1032 +0x5d
  gvisor.dev/gvisor/pkg/tcpip/adapters/gonet.(*TCPConn).Close()
      /home/mafredri/.local/go/pkg/mod/gvisor.dev/gvisor@v0.0.0-20221203005347-703fd9b7fbc0/pkg/tcpip/adapters/gonet/gonet.go:417 +0x4a
  github.com/coder/wgtunnel/tunneld.(*tracingConnWrapper).Close()
      /home/mafredri/src/coder/wgtunnel/tunneld/tracing.go:39 +0x7d
  net/http.(*persistConn).closeLocked()
      /usr/local/go/src/net/http/transport.go:2732 +0x222
  net/http.(*persistConn).readLoopPeekFailLocked()
      /usr/local/go/src/net/http/transport.go:2267 +0x344
  net/http.(*persistConn).readLoop()
      /usr/local/go/src/net/http/transport.go:2111 +0x1029
  net/http.(*Transport).dialConn.func5()
      /usr/local/go/src/net/http/transport.go:1765 +0x39

Goroutine 49358 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  github.com/coder/wgtunnel/tunneld_test.TestCompatibility()
      /home/mafredri/src/coder/wgtunnel/tunneld/tunneld_test.go:212 +0x449
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 50059 (finished) created at:
  net/http.(*Transport).dialConn()
      /usr/local/go/src/net/http/transport.go:1765 +0x2585
  net/http.(*Transport).dialConnFor()
      /usr/local/go/src/net/http/transport.go:1456 +0x13a
  net/http.(*Transport).queueForDial.func1()
      /usr/local/go/src/net/http/transport.go:1425 +0x47
==================
--- FAIL: TestCompatibility (0.00s)
    --- FAIL: TestCompatibility/Default (0.27s)
        testing.go:1446: race detected during execution of test
    --- FAIL: TestCompatibility/LongerPrefix (0.27s)
        testing.go:1446: race detected during execution of test
--- FAIL: TestEndToEnd (1.80s)
    testing.go:1446: race detected during execution of test
FAIL
FAIL	github.com/coder/wgtunnel/tunneld	22.358s
ok  	github.com/coder/wgtunnel/tunneld/httpmw	17.938s
FAIL
```

File: tunneld/api.go [hunk 3]
```diff
@@ -43,36 +48,123 @@ func (api *API) Router() chi.Router {
...
+func (api *API) registerClient(req tunnelsdk.ClientRegisterRequest) (tunnelsdk.ClientRegisterResponse, bool, error) {
+	if req.Version <= 0 || req.Version > tunnelsdk.TunnelVersionLatest {
+		req.Version = tunnelsdk.TunnelVersionLatest
+	}
+
+	ip, urls := api.WireguardPublicKeyToIPAndURLs(req.PublicKey, req.Version)
+
+	exists := false
 	if api.wgDevice.LookupPeer(req.PublicKey) == nil {
+		exists = true
```
[the remark this answers is not in the document]
Comment (deansheather) [reply tunneld/api.go:142]: Great catch!

File: tunneld/options.go (hunk 1 above)
Comment (deansheather) [reply tunneld/options.go:155]: You mean it's different to the old format generated by the current wgtunnel? It should be the same and there's tests for it...

File: tunneld/api.go (hunk 2 above)
Comment (deansheather) [reply tunneld/api.go:40]: Do you mean changing `/api/v1` to be `/api/v2`? The `/api/v1/clients` endpoint is the v2 version of the old `/tun` endpoint.

File: tunneld/options.go [hunk 4]
```diff
@@ -99,59 +100,106 @@ func (options *Options) Validate() error {
...
 }
 
-// WireguardPublicKeyToIP returns the IP address that corresponds to the given
-// wireguard public key.
-func (options *Options) WireguardPublicKeyToIP(publicKey device.NoisePublicKey) netip.Addr {
+// WireguardPublicKeyToIPAndURLs returns the IP address that corresponds to the
+// given wireguard public key, as well as all accepted tunnel URLs for the key.
+//
+// We support an older 32 character format ("old format") and a newer 12
+// character format ("good format") which is preferred. The first URL returned
```
[the remark this answers is not in the document]
Comment (deansheather) [reply tunneld/options.go:117]: Done, basically it's just because of the "good format" being shorter.

File: tunneld/options.go (hunk 1 above)
Comment (mafredri) [reply tunneld/options.go:157]: Ah, maybe just the way it was constructed changed then? I thought it looked different but maybe I'm blind, I trust you. 😄

Comment (deansheather) [reply tunneld/options.go:157]: I'll try it with a real current coder client post-merge and see what happens

**Pull request merged**